### PR TITLE
Paginated commits in tree view and version control sidebar

### PR DESCRIPTION
**Pull request by lukasbuenger**
###### Breaking
- The tree view now paginates over commits, therefore needs the paginated `Repo.commits` connection introduced in orbiting/backends#44
- In env-File: `API_URL` instead of `API_BASE_URL` and `API_WS_URL` instead of `API_WS_BASE_URL`
- Teasers get created with a  new key `data.id` to support orbiting/republik-frontend#126

###### Changes
- Updated to Apollo 2. 
- Apollo network policies changed to `cache-first` where sensible. 
- All dependencies updated except nextjs and slate. 
- Lots of additional null-checks in order to make `cache-first` work.

### CODE REVIEW

**Comment by tpreusse** [comment]
I noticed that you moved queries, mutations and fragments to a lib file each. I think this is an anti pattern. Moving the GraphQL code to a separate file is fine but query, mutation and fragments should stay close to the components using it, since they contain the fields needed to render. Additionally usually fragments are shared between mutations and queries dealing with the same type, ideally those could stay together in a file.

Also [React Apollo 2.1](https://www.apollographql.com/docs/react/react-apollo-migration.html#query-component) allows queries, mutations and subscription to be expressed as react components. Another reason to keep those things in the components folder.

**Comment by lukasbuenger** [comment]
This is WIP and not ready for review. 

I was aware of the fact, that this is not how one is supposed to write and file gql queries, but in order to understand what's going on and to get an understanding of what's redundant, I pulled off this setup. That's about to change again. 

While your usual Graph API might not care about a certain key getting fetched multiple times across different components, in our case it might lead to several fetches on the Github-API, which increases response time dramatically, especially if the Apollo Client has no way of caching the results. Example: `latestCommit` is equivalent to `nodes(first: 0)[0]` both in the API and on the server, still the caching mechanism has no way of knowing that.

File: lib/settings.js [hunk 1]
```diff
@@ -5,9 +5,9 @@ const ENV =
 
 export const LOCALE = ENV.LOCALE
 export const GITHUB_ORG = ENV.GITHUB_ORG
-export const API_BASE_URL = ENV.API_BASE_URL
+export const API_URL = ENV.API_URL
```
Comment (patrickvenetz) [inline lib/settings.js]: Would reflect environment variable naming improvement in `.env.example`

Comment (lukasbuenger) [reply lib/settings.js]: That's right, thank you!

**Comment by lukasbuenger** [comment]
Many thanks, @pae, that's very helpful indeed. 

Will fix unless stated otherwise.

File: pages/repo/tree.js [hunk 2]
```diff
@@ -88,24 +89,38 @@ class EditorPage extends Component {
...
+        document: treeRepoSubscription,
         variables: {
           repoId: this.props.url.query.repoId
         },
         updateQuery: (prev, { subscriptionData }) => {
           if (!subscriptionData.data) {
             return prev
           }
-          const { latestCommit, milestones } = subscriptionData.data.repoUpdate
+          const newLatestCommit = path(
```
[the remark this answers is not in the document]
Comment (patrickvenetz) [reply pages/repo/tree.js]: (On other hand, maybe not introduce Ramda complexity through the back door here.)

File: components/Frame/Me.js [hunk 3]
```diff
@@ -14,7 +14,7 @@ const styles = {
 
 export const Me = ({ me }) => (
   <div {...styles.container}>
-    {me.name || me.email}
+    {me && (me.name || me.email)}
     <br />
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply components/Frame/Me.js]: I don't know really. Will check it out. I really haven't touched the code unless necessary.

File: components/Repo/Table.js [hunk 4]
```diff
@@ -37,6 +38,88 @@ import EditMetaDate from './EditMetaDate'
...
+      productionDeadline
+      publishDate
+      briefingUrl
+    }
+  }
+}
+`
+
+export const filterAndOrderRepos = gql`
+query repoListSearch($after: String, $search: String, $orderBy: RepoOrderBy) {
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply components/Repo/Table.js]: This is holds for all your comments regarding the naming of GraphQL queries, mutations and subscriptions.

During development of this branch, I actually had all queries in one place (kind of, see https://github.com/orbiting/publikator-frontend/pull/149/commits/f54e3d5383cabcf6721fc9217d32b8a4572cbb19) to get an overview of what's there, what's redundant and especially, what is named how and why. 
I reverted the respective queries to be co-located with their dependants (see comment from @tpreusse above). 

I do very much agree on the fact, that we have to come up with a couple of sensible naming conventions, but this is not something I'm willing to do on my own without having an overall agreement on how to approach this.

File: components/Tree/index.js [hunk 5]
```diff
@@ -331,12 +351,13 @@ class Tree extends Component {
...
-          commits.map(commit =>
-            <span
+          commits.map(commit => {
+            const colors = this.getColor(commit.author.email)
+            return <span
               key={commit.id}
               ref={commit.setNodeRef}
               style={{
-                backgroundColor: commit.color
+                backgroundColor: colors.color
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply components/Tree/index.js]: Will do.

File: components/VersionControl/UncommittedChanges.js [hunk 6]
```diff
@@ -22,6 +23,37 @@ import createDebug from 'debug'
...
+      uncommittedChanges {
+        id
+        email
+        name
+      }
+    }
+  }
+`
+
+export const uncommittedChangesSubscription = gql`
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply components/VersionControl/UncommittedChanges.js]: See above

File: components/VersionControl/index.js [hunk 7]
```diff
@@ -78,12 +76,18 @@ class EditSidebar extends Component {
             return prev
           }
           const { commits } = subscriptionData.data.repoUpdate
-          if (commits && commits.length) {
+          if (commits && commits.nodes.length) {
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply components/VersionControl/index.js]: No. See signature of query `Repo.commits` and its resulting `CommitConnection`

File: components/editor/modules/embed/index.js [hunk 8]
```diff
@@ -9,7 +9,9 @@ import createUi from './ui'
 import embedFromUrlPlugin from './embedFromUrlPlugin'
 import EmbedLoader from './EmbedLoader'
 
-const videoQuery = gql`
+import gql from 'graphql-tag'
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply components/editor/modules/embed/index.js]: Will do

File: components/editor/utils/RepoSearch.js [hunk 9]
```diff
@@ -1,12 +1,13 @@
...
+import { graphql } from 'react-apollo'
+import gql from 'graphql-tag'
 import { Autocomplete, InlineSpinner } from '@project-r/styleguide'
 import debounce from 'lodash.debounce'
 
 import { GITHUB_ORG, REPO_PREFIX } from '../../../lib/settings'
 
-const repoQuery = gql`
-query repos($after: String, $search: String) {
+export const filterRepos = gql`
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply components/editor/utils/RepoSearch.js]: See above.

File: pages/repo/edit.js [hunk 10]
```diff
@@ -122,22 +57,76 @@ const commitMutation = gql`
...
 
-const uncommittedChangesMutation = gql`
+export const uncommittedChangesMutation = gql`
   mutation uncommittedChanges($repoId: ID!, $action: Action!) {
     uncommittedChanges(repoId: $repoId, action: $action)
   }
 `
 
+export const getCommitById = gql`
+    query getCommitById($repoId: ID!, $commitId: ID!) {
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply pages/repo/edit.js]: Will do.

File: pages/repo/edit.js [hunk 11]
```diff
@@ -728,30 +717,39 @@ export class EditorPage extends Component {
...
-                {!readOnly && <Sidebar.Tab tabId='edit' label='Editieren'>
-                  {!!this.editor && <EditorUI
+            </div>
+          )} />
+          <Sidebar warnings={warnings}
+            isDisabled={Boolean(showLoading || error)}
+            selectedTabId={(readOnly && 'workflow') || undefined}
+            isOpen={showSidebar}
+          >
+
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply pages/repo/edit.js]: Will do.

File: pages/repo/edit.js [hunk 12]
```diff
@@ -843,13 +841,22 @@ export default compose(
...
               data
             })
-          }
+          },
+          refetchQueries: [
+            {
+              query: getRepoHistory,
+              variables: {
+                repoId: url.query.repoId,
+                first: 10
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply pages/repo/edit.js]: Good catch. Will do.

File: pages/repo/tree.js [hunk 13]
```diff
@@ -9,73 +13,70 @@ import Loader from '../../components/Loader'
...
-        name
-      }
-    }
-  `
-}
-
-export const query = gql`
-  query repoWithHistory($repoId: ID!) {
+export const getRepoHistory = gql`
+  query repoWithHistory(
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply pages/repo/tree.js]: See above.

File: pages/repo/tree.js [hunk 14]
```diff
@@ -9,73 +13,70 @@ import Loader from '../../components/Loader'
...
-  ${fragments.commit}
-  ${fragments.milestone}
+  ${fragments.SimpleMilestone}
+  ${fragments.SimpleCommit}
 `
 
-const repoSubscription = gql`
-  subscription repoUpdate($repoId: ID!) {
+export const treeRepoSubscription = gql`
+  subscription onRepoUpdate($repoId: ID!) {
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply pages/repo/tree.js]: See above.

File: pages/repo/tree.js [hunk 15]
```diff
@@ -173,12 +199,52 @@ class EditorPage extends Component {
...
+          first: 20
+        },
+        notifyOnNetworkStatusChange: true,
+        fetchPolicy: 'cache-and-network'
+      })
+    },
+    props: ({data, ownProps}) => {
+      return ({
+        data,
+        commits: (data.repo && data.repo.commits && data.repo.commits.nodes) || {},
```
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply pages/repo/tree.js]: True. Will do.

File: pages/repo/tree.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukasbuenger) [reply pages/repo/tree.js]: I mean, I'd gladly use Ramda all over the place, but it really adds a lot of complexity for people not familiar with Ramda and FP principles in general. 

And this is, again, not something that I'm willing to decide without the consent of the whole team.

**Pull request merged**